File: src/main/java/com/github/cao/awa/lunaria/Lunaria.java
package com.github.cao.awa.lunaria;

import com.github.cao.awa.lunaria.status.LunariaStatus;

import java.util.function.Consumer;

/**
 * The abstract base class for Lunaria framework operations.
 * This class provides core functionality for managing virtual thread-based operations
 * with state tracking and exception handling capabilities.
 *
 * @author cao_awa
 *
 * @since 1.0.0
 */
abstract public class Lunaria {
    /** The virtual thread that executes the Lunaria operation */
    private Thread thread;
    /** Flag indicating whether the operation is completed */
    public boolean isDone;
    /** Current status of the Lunaria operation */
    private LunariaStatus status = LunariaStatus.RUNNING;
    private Throwable exception;
    private Consumer<Throwable> exceptionHandler;

    public Lunaria() {
        // Initialize the virtual thread with the action provided by getAction()
        this.thread = Thread.ofVirtual().start(getAction());
    }

    public LunariaStatus getStatus() {
        return this.status;
    }

    /**
     * Abstract method that must be implemented to provide the core operation logic.
     *
     * @return A Runnable containing the operation to be executed
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    public abstract Runnable getAction();

    /**
     * Handles any exception that occurred during execution.
     * Invokes the exception handler and updates the operation state to FAILED.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    public void requestHandleException() {
        if (this.exception != null && this.exceptionHandler != null) {
            this.exceptionHandler.accept(this.exception);
        }
        this.status = LunariaStatus.FAILED;
    }

    public void exceptionHappening(Throwable ex) {
        this.exception = ex;
        requestHandleException();
    }

    /**
     * Cancels the current operation.
     * Attempts to interrupt the thread and updates the operation state.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    public void cancel() {
        try {
            this.thread.interrupt();
            markDone();
            this.status = LunariaStatus.CANCELLED;
            return;
        } catch (Throwable ex) {
            this.exception = ex;
        }
        markDone();
        this.status = LunariaStatus.FAILED;
    }

    /**
     * Marks the operation as completed successfully.
     * Updates the operation state to DONE.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    public void markDone()  {
        this.isDone = true;
        this.status = LunariaStatus.DONE;
    }

    public void setSetExceptionHandler(Consumer<Throwable> handler) {
        this.exceptionHandler = handler;
    }
}

File: src/main/kotlin/com/github/cao/awa/lunaria/Lunaria.kt
package com.github.cao.awa.lunaria

import com.github.cao.awa.lunaria.consumer.ConsumerLunaria
import com.github.cao.awa.lunaria.state.LunariaState
import com.github.cao.awa.lunaria.supplier.SupplierLunaria

/**
 * The abstract base class for Lunaria framework operations.
 * This class provides core functionality for managing virtual thread-based operations
 * with state tracking and exception handling capabilities.
 *
 * @author cao_awa
 *
 * @since 1.0.0
 */
abstract class Lunaria {
    /** The virtual thread that executes the Lunaria operation */
    private var thread: Thread
    /** Flag indicating whether the operation is completed */
    var isDone: Boolean = false
    /** Current state of the Lunaria operation */
    var state: LunariaState = LunariaState.RUNNING
    /** Stores any exception that occurred during execution */
    var exception: Throwable? = null
    /** Handler for processing exceptions that occur during execution */
    var exceptionHandler: ((Throwable) -> Unit) = { }

    init {
        // Initialize the virtual thread with the action provided by getAction()
        this.thread = Thread.ofVirtual().start(getAction())
    }

    /**
     * Abstract method that must be implemented to provide the core operation logic.
     *
     * @return A Runnable containing the operation to be executed
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    abstract fun getAction(): Runnable

    /**
     * Handles any exception that occurred during execution.
     * Invokes the exception handler and updates the operation state to FAILED.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    fun handleException() {
        if (this.exception != null) {
            this.exceptionHandler(this.exception!!)
        }
        this.state = LunariaState.FAILED
    }

    /**
     * Cancels the current operation.
     * Attempts to interrupt the thread and updates the operation state.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    fun cancel() {
        this.exception = runCatching {
            this.thread.interrupt()
        }.exceptionOrNull()
        this.isDone = true
        this.state = LunariaState.CANCELLED
    }

    /**
     * Marks the operation as completed successfully.
     * Updates the operation state to DONE.
     *
     * @author cao_awa
     *
     * @since 1.0.0
     */
    fun markDone() {
        this.isDone = true
        this.state = LunariaState.DONE
    }
}